hazelcast-client-legacy/src/main/java/com/hazelcast/client/spi/impl/listener/LazyDistributedObjectEvent.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2008-2015, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.spi.impl.listener;
+
+import com.hazelcast.client.spi.ProxyManager;
+import com.hazelcast.core.DistributedObject;
+import com.hazelcast.core.DistributedObjectEvent;
+
+public class LazyDistributedObjectEvent extends DistributedObjectEvent {
+
+    private String objectName;
+    private ProxyManager proxyManager;
+
+    /**
+     * Constructs a DistributedObject Event.
+     *
+     * @param eventType         The event type as an enum {@link EventType} integer.
+     * @param serviceName       The service name of the DistributedObject.
+     * @param distributedObject The DistributedObject for the event.
+     * @param proxyManager      The ProxyManager for lazily creating the proxy if not available on the client.
+     */
+    public LazyDistributedObjectEvent(EventType eventType, String serviceName, String objectName,
+                                      DistributedObject distributedObject,
+                                      ProxyManager proxyManager) {
+        super(eventType, serviceName, distributedObject);
+        this.objectName = objectName;
+        this.proxyManager = proxyManager;
+    }
+
+    @Override
+    public DistributedObject getDistributedObject() {
+        if (EventType.DESTROYED.equals(eventType)) {
+            return null;
```

Comment:
have you considered throwing an exception here? in every case the behaviour should be explicitly documented in contract.